### DISCUSSION TITLE
Discussion: Relax type requirements for ??

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#1720] [General]
As part of the implementation for `??=` (championed issue [here](https://github.com/dotnet/csharplang/issues/34)), we're looking at relaxing the requirement on the type of `A` in `??` (where `A` is the type of `a` in `a ?? b`). Currently, the spec for `??` states:

> 1. If A exists and is not a nullable type or a reference type, a compile-time error occurs.

We're planning on relaxing this constraint to be:

> 1. If A exists and is a non-nullable value type, a compile-time error occurs.

This will enable `t1 ?? t2`, where the type of `t1` and `t2` are an unconstrained generic parameter, which is currently illegal. We will also enable this for `??=`, which will have a similar constraint (for `??=`, `A` must exist).

**Comment by orthoxerox** [comment]
See also #505, #263

**Comment by franklin-ross** [comment]
I would also love to see the same relaxation for `t1 is null` where t1 is an unconstrained generic parameter. It's currently not obvious how to do that effectively and efficiently IMO. The only way I know of is `!(value is T)`, and that looks so strange. Some of the other ways I've seen either likely cause boxing, are not the same, and/or go through the equality functions. Things like `(object)default(T) is null`, `(object)t1 == null`, and code fixes like `EqualityComparer<T>.Default.Equals(value, default(T))`.

Here's a [discussion I had over at Roslynator about it](https://github.com/JosefPihrt/Roslynator/issues/384). A web search brings up a fair few [others like this](https://stackoverflow.com/questions/12396457/checking-instance-of-non-class-constrained-type-parameter-for-null-in-generic-me).

`t1 is null` very concisely says what it means, avoids boxing, and avoids equality machinery.

**Comment by marksmeltzer** [comment]
All the things, please 😁

**Comment by 333fred** [comment]
@franklin-ross could you open a separate issue for that? This issue is tracking relaxation for `??`. That being said, I think it would be likely @gafter or someone else from the LDM would champion that.

**Comment by franklin-ross** [comment]
@333fred Yup, sure. Sorry if that was a bit random here, it just seemed related 😅

**Comment by gafter** [comment]
@franklin-ross We're relaxing things to allow `t1 is null` in C# 8.0.  See https://github.com/dotnet/csharplang/issues/1284.  It is already implemented in the preview language version using VS 2019.

**Comment by franklin-ross** [comment]
I was very pleased to see that when I went to create a new issue, @gafter. Thanks for posting the link here for posterity.
